libs/ml/include/ml/core/node.hpp

```diff
@@ -141,18 +154,184 @@ class Node
     return static_cast<bool>(cached_output_status_ == CachedOutputState::VALID_CACHE);
   }
 
+  /**
+   * @brief ForwardPassChargeCost calculates charge cost estimation of the current node.
+   * The cost includes also charge costs for every preceding Node in a linear (sequential)
+   * Graph up to input Node.
+   * @return charge cost estimation for all Forward passes through all Nodes starting from input
+   * and ending in this node.
+   */
+  // TODO(VH): Naming does not suggest recursive nature of this method, renaming needed.
+  //           PrecedingNodesForwardCost or just PrecedingForwardCost?
+  virtual fetch::vm::ChargeAmount ForwardPassChargeCost()
+  {
+    // TODO(VH): call OpForwardCost() of the Op - we need to know Data shape to do this.
+    if (input_nodes_.empty())
+    {
+      auto const &expected_in_shapes = op_ptr_->ExpectedSliceInputShapes();
+      if (this->OpCode() == OpType::OP_PLACEHOLDER)
+      {
+        FETCH_LOG_INFO(name_.c_str(), "Input Placeholder layer cost added");
+      }
+      else
+      {
+        FETCH_LOG_ERROR(name_.c_str(),
+                        "Can not estimate a forward pass cost of a non-placeholder leaf Node!");
+      }
+      return op_ptr_->OpForwardCost(expected_in_shapes);
```

Comment:
I think static analysis wont permit this anyway; you'll need to do something like:
if (input_nodese_.empty())
{
  ...
  ret = op_ptr_->OpForwardCost(expected_in_shapes);
}
else
{
  ret = ...;
}
return ret;